### PR TITLE
add dirs and detect logic for custom python packs

### PR DESCRIPTION
**Pull request by ranginuitrot**

### CODE REVIEW

**Comment by ranginuitrot** [comment]
Note: detect logic is in `detect` file in each pack dir.

**Comment by ranginuitrot** [comment]
@yyolk definitely not ready yet, but please put comments here if you want to test them.

File: packs/python-model-microservice/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        branch 'master'
+      }
+      steps {
+        container('python') {
+          // ensure we're not on a detached head
+          sh "git checkout master"
+          sh "git config --global credential.helper store"
+          sh "jx step git credentials"
+          // so we can retrieve the version in later steps
+          sh "echo \$(jx-release-version) > VERSION"
```
Comment (yyolk) [inline packs/python-model-microservice/Jenkinsfile]: you were mentioning that this was tracked via setup.py? 

sorry for the ignorant question I found https://github.com/jenkins-x/jx-release-version and didn't know how it works with this buildpack or is this only for like the scope of the build?

File: packs/python-model-microservice/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  agent {
+    // also available: "jenkins-python2-gpu"
+    label "jenkins-python-gpu"
+  }
+  environment {
+    ORG = 'REPLACE_ME_ORG'
+    APP_NAME = 'REPLACE_ME_APP_NAME'
+    CHARTMUSEUM_CREDS = credentials('jenkins-x-chartmuseum')
+    HELM_RELEASE = "$PREVIEW_NAMESPACE".toLowerCase()
+    PREVIEW_VERSION = "0.0.0-SNAPSHOT-$BRANCH_NAME-$BUILD_NUMBER"
```
Comment (yyolk) [inline packs/python-model-microservice/Jenkinsfile]: another ignorant question - should this align with the "app version"?

File: packs/python-model-microservice/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,135 @@
+pipeline {
+  agent {
+    // also available: "jenkins-python2-gpu"
+    label "jenkins-python-gpu"
+  }
+  environment {
+    ORG = 'REPLACE_ME_ORG'
+    APP_NAME = 'REPLACE_ME_APP_NAME'
+    CHARTMUSEUM_CREDS = credentials('jenkins-x-chartmuseum')
+    HELM_RELEASE = "$PREVIEW_NAMESPACE".toLowerCase()
```
Comment (yyolk) [inline packs/python-model-microservice/Jenkinsfile]: move below `PREVIEW_NAMESPACE = ...`

File: packs/python-library/detect [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
+BUILD_DIR=$1
+
+# Exit early if app is clearly not a Python Model Microservice.
+if [ ! -f $BUILD_DIR/Makefile ] || [ $(! grep -c ^project-type: $BUILD_DIR/Makefile) ] || [ "python-library" != $(make -f $BUILD_DIR/Makefile -s project-type) ]; then
+  exit 1
+fi
+
+echo Python-Model-Microservice
```
Comment (yyolk) [inline packs/python-library/detect]: 👍

File: packs/python-model-microservice/skaffold.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+apiVersion: skaffold/v1alpha2
+kind: Config
+build:
+  tagPolicy:
+    envTemplate:
+      template: "{{.DOCKER_REGISTRY}}/REPLACE_ME_ORG/REPLACE_ME_APP_NAME:{{.VERSION}}"
+  artifacts:
+  - imageName: changeme
```
Comment (yyolk) [inline packs/python-model-microservice/skaffold.yaml]: is this the right value?

File: packs/python/detect [hunk 6]
```diff
@@ -7,4 +7,9 @@ if [ ! -f $BUILD_DIR/requirements.txt ] && [ ! -f $BUILD_DIR/setup.py ] && [ ! -
   exit 1
 fi
 
+# Exist early if app is a custom Python project.
```
[the remark this answers is not in the document]
Comment (ranginuitrot) [reply packs/python/detect]: ✅

File: packs/python-model-microservice/Jenkinsfile (hunk 1 above)
Comment (ranginuitrot) [reply packs/python-model-microservice/Jenkinsfile]: It's extracted from Chart.yaml's version. So this lets us track deployment and library versions separately. Not sure if this is the right path though.

File: packs/python-model-microservice/Jenkinsfile (hunk 2 above)
Comment (ranginuitrot) [reply packs/python-model-microservice/Jenkinsfile]: I'm not sure what it should match. So far, it looks like it's ignored in favor of Chart.yaml's version.

File: packs/python-model-microservice/Jenkinsfile (hunk 3 above)
Comment (ranginuitrot) [reply packs/python-model-microservice/Jenkinsfile]: ✅

File: packs/python-model-microservice/skaffold.yaml (hunk 5 above)
Comment (ranginuitrot) [reply packs/python-model-microservice/skaffold.yaml]: Not sure yet. Rather, I'm not sure what effect this has.

File: packs/python-model-microservice/Jenkinsfile (hunk 1 above)
Comment (yyolk) [reply packs/python-model-microservice/Jenkinsfile]: 👍

File: packs/python-model-microservice/skaffold.yaml (hunk 5 above)
Comment (yyolk) [reply packs/python-model-microservice/skaffold.yaml]: Gotcha. Guess we'll find out.

File: packs/python-model-microservice/Jenkinsfile (hunk 2 above)
Comment (yyolk) [reply packs/python-model-microservice/Jenkinsfile]: Cool, thanks!

**Comment by ranginuitrot** [comment]
The three areas to look at are `packs/python-library|python-model-microservice|docker`. Probably easiest to browser files directly.

**Comment by ranginuitrot** [comment]
@yyolk it's ready for you :)

File: packs/python-library/Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            base_version = sh(script: "pipenv run setup.py --version", returnStdout: true).trim()
+            pr = BRANCH_NAME.replaceAll(/-/, "").toLowerCase()
+            expanded_version = "${base_version}+${pr}.${BUILD_NUMBER}"
+
+            sh "sed -i -e 's/^__version__ = .*/__version__ = \"$expanded_version\"/' setup.py"
+
+            sh "pipenv run setup.py sdist bdist_wheel"
+            sh "twine upload --repository-url ${TWINE_REPO} dist/*"
+
+            pullRequest.comment("Successfully built and published **$expanded_version**")
```
[the remark this answers is not in the document]
Comment (ranginuitrot) [reply packs/python-library/Jenkinsfile]: turns out `jx step pr comment` can only comment on github issues—not PRs (don't ask me why `pr` is in the command...).

File: packs/python-model-microservice/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    curl \
+    libfreetype6-dev \
+    libpng12-dev \
+    libzmq3-dev \
+    pkg-config \
+    rsync \
+    unzip
+
+RUN apt-get clean && rm -rf /var/lib/apt/lists/*
```
[the remark this answers is not in the document]
Comment (ranginuitrot) [reply packs/python-model-microservice/Dockerfile]: Changed in [odfc610](https://github.com/AmFamLabs/draft-packs/pull/1/commits/0dfc61075ffd75326c018949fac895c05e76904c).

File: packs/python-model-microservice/Jenkinsfile [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    }
+    stage('Smoketest') {
+      steps {
+        container('python') {
+          script {
+            run_smoketest = sh(returnStatus: true, script: "grep -c ^${SMOKETEST_CMD}: Makefile")
+            if (run_smoketest == 0) {
+              repo = BRANCH_NAME == 'master' ? "$DOCKER_REGISTRY/$ORG/$APP_NAME:${sh(script: 'cat VERSION', returnStdout: true)}" : "$DOCKER_REGISTRY/$ORG/$APP_NAME:${sh(script: 'cat PREVIEW_VERSION', returnStdout: true)}"
+              hostname = BRANCH_NAME == 'master' ? "$APP_NAME.$STAGING_NAMESPACE.$DOMAIN_POSTFIX" : "$APP_NAME.$PREVIEW_NAMESPACE.$DOMAIN_POSTFIX"
+              sh "docker run --entrypoint=/bin/sh -e PIP_EXTRA_INDEX_URL=$PIP_EXTRA_INDEX_URL ${repo} sh -c 'make $SMOKETEST_CMD SMOKE_HOST=${hostname} BRANCH=$BRANCH_NAME'"
```
[the remark this answers is not in the document]
Comment (ranginuitrot) [reply packs/python-model-microservice/Jenkinsfile]: Agreed on having a dockerfile-defined ENTRYPOINT. Changed in [3c83d7b](https://github.com/AmFamLabs/draft-packs/commit/3c83d7b10230b786d9bcdf98d23b8b4e36c89fbb).

**Comment by ranginuitrot** [comment]
@yyolk please have a look at my latest in regards to your above unresolved conversations.

**Pull request merged**